stream_alert/shared/lookup_tables/driver_s3.py

```diff
@@ -0,0 +1,195 @@
+from stream_alert.shared.lookup_tables.cache import DriverCache
+from stream_alert.shared.lookup_tables.drivers import PersistenceDriver
+from stream_alert.shared.lookup_tables.errors import LookupTablesInitializationError
+
+LOGGER = get_logger(__name__)
+
+
+class S3Driver(PersistenceDriver):
+    """
+    S3Driver
+
+    This PersistenceDriver uses AWS S3 as the backing layer. The entire table is stored in a single
+    large S3 file.
+
+    Upon initialization, the S3 file is loaded, decompressed, decoded, and loaded into memory as
+    a Python dict.
+
+    To improve performance, the decoded contents of the S3 file are cached in-memory for up to
+    "cache_refresh_minutes" in minutes of time, after which they are reloaded from S3.
+
+    (!) NOTE: Tweaking the cache_refresh_minutes will likely not help memory performance of your
+        LookupTable. If you are having memory issues on your Lambda, try using the DynamoDBDriver.
+    """
+    _DEFAULT_CACHE_REFRESH_MINUTES = 10
+    _MAGIC_CACHE_TTL_KEY = '____LAST_LOAD_TIME____'
+
+    def __init__(self, configuration):
+        # Example configuration:
+        # {
+        #     "driver": "s3",
+        #     "bucket": "airbnb.sample.lookuptable",
+        #     "key": "resource_map.gz",
+        #     "cache_refresh_minutes": 10,
+        #     "compression": "gzip",
+        #     "key_delimiter": "|"
+        # },
+
+        super(S3Driver, self).__init__(configuration)
+
+        self._s3_bucket = configuration['bucket']
+        self._s3_key = configuration['key']
+        self._compression = configuration.get('compression', False)
+        self._cache_refresh_minutes = configuration.get(
+            'cache_refresh_minutes',
+            self._DEFAULT_CACHE_REFRESH_MINUTES
+        )
+
+        self._cache = DriverCache(maximum_key_count=0)
+
+        self._dirty = False
+
+        # Explicitly set timeout for S3 connection. The boto default timeout is 60 seconds.
+        boto_config = boto_helpers.default_config(timeout=10)
+        self._s3_client = boto3.resource('s3', config=boto_config)
+
+    @property
+    def driver_type(self):
+        return self.TYPE_S3
+
+    @property
+    def id(self):
+        return '{}:{}/{}'.format(self.driver_type, self._s3_bucket, self._s3_key)
+
+    def initialize(self):
+        LOGGER.info('LookupTable (%s): Running initialization routine', self.id)
+        self._reload()
+
+    def commit(self):
+        if not self._dirty:
+            LOGGER.warn('LookupTable (%s): Empty commit; no records dirtied', self.id)
+            return
+
+        raise NotImplementedError('Help derek you screwed me')
+
+    def get(self, key, default=None):
+        self._reload_if_necessary()
+        return self._cache.get(key, default)
+
+    def set(self, key, value):
+        """
+        Under construction
+        FIXME (derek.wang)
+        """
+        self._dirty = True
+
+    def _reload_if_necessary(self):
```

Comment:
This method is almost same in DynamoDB driver class. Can we move the common method(s) to base class `PersistenceDriver`?